File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {

        unordered_set<char>st;

        int l=0;
        // int r=s.size()-1;
        int maxL=0;

        for(int i=0;i<s.size();i++)
        {
            while(st.count(s[i]))
            {
                st.erase(s[l]);
                l++;
            }

            maxL=max(i-l+1,maxL);

            st.insert(s[i]);
        }

        return maxL;
        
    }
};

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        window=set()
        maxL=0

        left=0

        for right in range(len(s)):
            curr_c=s[right]

            
            while curr_c in window:
                window.remove(s[left])
                left+=1

            maxL=max(maxL, right-left+1)

            window.add(curr_c)
        return maxL
# TC=O(n)
# SC=O(n)